fenced/utils.py

```diff
@@ -1,169 +1,61 @@
+            not ignore.pattern.match(entry.name),
+            entry.name not in ignore.user_excluded_disks,
+        )
+    )
+
+
+def load_disks_from_sys_block(ignore: IgnoreObject) -> dict[str, str]:
+    """By the time fenced is called, the OS is booted into multi-user
+    mode and has been for a bit. This means we should use sysfs to
+    enumerate the disks since using things like libudev, are fickle
+    and can have missing devices based on operations done in userspace.
+    (i.e. if someone causes a "rescan"). Sysfs does not suffer from
+    such scenarios. (Obviously, sysfs can be "updated" if the drive
+    catastrophically fails OR someone physically pulls it or someone
+    issues an hba reset, etc but there isn't a way to prevent such
+    situations in the first place)"""
     disks = {}
     try:
-        stdout = subprocess.run(cmd, stdout=subprocess.PIPE).stdout.decode()
-        for i in json.loads(stdout)["blockdevices"]:
-            if i["name"].startswith(ignore[0]) or ignore[1].match(i["name"]):
-                continue
-            else:
-                disks[i["name"]] = i["serial"]
+        with scandir("/sys/block") as sdir:
+            for disk in filter(lambda x: should_not_ignore(x, ignore), sdir):
+                disks[disk.name] = disk.name
     except Exception:
-        logger.error("Unhandled exception", exc_info=True)
+        logger.error("Unhandled exception enumerating disks", exc_info=True)
 
     return disks
 
 
-def disks_to_be_ignored(ed=None):
-    prefixes = {"sr", "md", "dm-", "loop", "zd", "pmem"}
-    if ed is not None:
-        exclude = {}
-        try:
-            if isinstance(ed, list):
-                # it's a list by default if fenced is called without exclude disk args
-                exclude = {i for i in ed}
-            elif isinstance(ed, str):
-                # it's a comma separated list if fenced is called with exclude disk args
-                exclude = {i for i in ed.split(",") if i}
-        except Exception:
-            logger.warning("Failed to format exclude disks params", exc_info=True)
-        else:
-            if exclude:
-                prefixes.update(exclude)
-
-    return (tuple(prefixes), re.compile(r"nvme[0-9]+c"))
-
-
-def load_disks_impl(exclude_disks=None, use_zpools=False):
+def load_disks_impl(ed: tuple[str] | tuple, use_zpools: bool = False) -> dict[str, str]:
     """
-    The name of the game for the various functions are ultimate fault tolerance.
-    Since fenced is paramount for preventing zpool corruption, we do everything we
-    can to prevent unexpected failures and always return disks.
+    Return disk(s) to have persistent reservations placed upon.
+        `ed`: user supplied disk(s) that will be excluded from having
+            persistent reservations placed upon them.
+        `use_zpools`: boolean is a NO_OP (for now) and is a placeholder
+            for generating a group of disks ONLY associated to a zpool.
```

Comment:
No longer see parameter `use_zpools` being used.  Remove it?

Reply:
It's used in middleware repo so kept it here to reduce code churn and prevent risk of crashing with too many args error.